### ISSUE TITLE
Move CUDA Low Level API to other namespace

### ISSUE THREAD

**Issue by okuta** [cupy/cupy#3385]
The `cupy.cuda` namespace is confusing.

1. `cupy.cuda` is not in NumPy
2.  Mixing Low Level API and CuPy original wrapper

Especially solving the second problem simplifies CuPy development.
(The first problem is a bit difficult because it requires design discussion.)

The implementation of some features on CuPy does not provide CUDA's Low Level API. Therefore, I think it is better to start with partial separation first.

With this change, CuPy may be able to provide Cython CUDA API in the future #130.

**Comment by leofang** [comment]
I have mixed feeling about the proposed change. 

First, I hope it's backward compatible (looks like so in #3386?). A lot of our code and a few downstream libraries that I know assume the existence of `cupy.cuda`. The runtime API is also documented. 

I view the fact that NumPy does not have `numpy.cuda` as an upside, and don't feel the obligation that CuPy must not have it for compatibility reason. It is obvious for GPU users that a `cuda` module must exist somewhere, so it may as well stay right under the main module.

Second, not directly related to this thread, but this proposal could potentially enable a separate module like `cupy.rocm` (or `cupy_rocm`) for AMD users. I understand that HIP is basically a thin layer on top of CUDA API, but it might be useful to make a distinction (at the level of user code; for the underlying implementation it's probably easier to stay with the current header-replacement approach)? This requires further discussion, though.

**Comment by okuta** [comment]
> First, I hope it's backward compatible (looks like so in #3386?). A lot of our code and a few downstream libraries that I know assume the existence of cupy.cuda. The runtime API is also documented.
Yes, I want to keep backward compatible. 

Stream handling in the Low Level API may need to be changed. The current low level API uses the current steam from Python thread local. It may be better to separate steam operation from the API.


> Second, not directly related to this thread, but this proposal could potentially enable a separate module like cupy.rocm (or cupy_rocm) for AMD users.
Yes, I agree.

**Comment by leofang** [comment]
Hi @okuta One more question: Would this help resolve the complex circular dependency between `cupy.cuda` and `cupy.core`? Ideally the former should be self-contained and be a dependency of the latter.

**Comment by okuta** [comment]
I agree with you.

We are currently experiencing a dependency from a Stream issue.
A Stream dependency now exists. I will send a PR that resolve it.

**Comment by leofang** [comment]
Two more comments (sorry for keeping bugging you @okuta 😅 I just think this is the perfect time and place to bring up everything related):
1. Once this is resolved, am I understanding it correctly that `cupy.cudnn`, `cupy.cusparse`, etc, which are more like CuPy's original wrappers than low-level APIs, will be moved back to under `cupy.cuda`? When I first realized their existence it took me a while to figure out what's going on.
2. There was a discussion about how CUDA low-level APIs can be "officially" exposed to the Python space (see https://github.com/numba/numba/issues/4797; officially in the sense that all Python CUDA programs can -- and better -- depend on this package), and there are a few suggestions, one of which being our `cupy_cuda`-to-be. In fact, this was brought up because of the need to convert a CuPy stream to a Numba one. If you're interested in making `cupy_cuda` a standalone package, this would be the long sought-after, ultimate "singleton" solution. 
I understand there will be a maintenance challenge no doubt, but given the resource in PFN/NVIDIA perhaps it's worth considering? I'll be more than happy to help maintain if needed 🙂 

cc: @seibert @gmarkall for the 2nd issue above.

**Comment by okuta** [comment]
Thanks @leofang .
> Once this is resolved, am I understanding it correctly that cupy.cudnn, cupy.cusparse, etc, which are more like CuPy's original wrappers than low-level APIs, will be moved back to under cupy.cuda? When I first realized their existence it took me a while to figure out what's going on.

I think it is necessary to move CuPy original wrappers (`cupy.cudnn`, `cupy.cusparse` ). But, I don't have good idea. There are `cupy.cuda.cudnn` and `cupy.cuda.cusparse`. For example, we may need a new namespace...

> There was a discussion about how CUDA low-level APIs can be "officially" exposed to the Python space (see numba/numba#4797; officially in the sense that all Python CUDA programs can -- and better -- depend on this package), and there are a few suggestions, one of which being our cupy_cuda-to-be. In fact, this was brought up because of the need to convert a CuPy stream to a Numba one. If you're interested in making cupy_cuda a standalone package, this would be the long sought-after, ultimate "singleton" solution.
I understand there will be a maintenance challenge no doubt, but given the resource in PFN/NVIDIA perhaps it's worth considering? I'll be more than happy to help maintain if needed 🙂

I hope it. This separation gives CuPy two possibilities.
- Easy replacement of CuPy backend
- It is possible that the `cupy_cuda` namespace can provide a common one as a CUDA Cython interface.

However, standardizing the interface is a difficult task. So I think it's best to work step by step.
This separation is first step of standard

**Comment by okuta** [comment]
> The cupy.cuda namespace is confusing.
> 1. cupy.cuda is not in NumPy
> 2. Mixing Low Level API and CuPy original wrapper

These were the problems I wanted to solve. However, I gave up moving the whole `cupy.cuda` because `cupy.cuda` has a dependency on `cupy.core.ndarray`. A circular reference exists when `cupy.cuda` is moved to another namespace (e.g., `cupy_XXXX`).


Therefore, I introduced `cupy_backends.cuda.api` and `backends.cuda.libs`. I moved low-level  CUDA API wrappers.
[The problem of the Stream was avoided by introducing a flag.](https://github.com/cupy/cupy/blob/fcd1761eb5cf2b593bd353005b89284478c03fe0/cupy_backends/cuda/stream.pxd#L3)


`cupy_backends.cuda` is a useful namespace. Wouldn't it be a good idea to cut out features that could be used in other libraries? #3584 is an example implementation of this.

For example, I think it's good to put the following items in `cupy_backends.cuda`.
- Things that have life cycle and create-destroy method. For example, Stream/Event/Memory etc.
- Thin Pythonic wrappers. Like this.

- Lifecycle of create-destroy. For example, Stream/Event/Memory etc.
Thin Pythonic wrappers. [Like this.](https://github.com/NVIDIA/pynvrtc/tree/master/pynvrtc 
)

I'd like to get your opinion.
I'm going to make a different issue because it's not on the same topic as the original.
